README.md
gPowerSwitch
============

gPowerSwitch is a GTK+ based GUI (graphical user interface) for [Objective Development's](http://obdev.at) [PowerSwitch](http://www.obdev.at/products/vusb/powerswitch.html) - USB driven control switch. Program can be helpful when you don't like command line tools or when you just need a simple GUI for PowerSwitch. But remember, gPowerSwitch is only a PC application, for hardware project please go to OBDev's page.

Required libraries:
* GTK+ >= 2.6
* LibGLADE >= 2.0
* LibUSB


## Configuration and compilation ###

The configuration of Vendor and Device names can be done in `configure.in`
After change it you have to `./autogen.sh` the project
WARNING: It's must be the same as in the firmware!

In powerSwitch.h you can change the `PSG_ACTIVE` macro, I've defined it for easier
change of active state, because on my EvBoard it's zero..


NEWS
2007-01-12 - 0.3.0

	- no .c files generated by Glade, use LibGLADE for GUI creation
	- no aboutdialog in .glade file, dialog is generated on-the-fly after
	  'About' button click
	- bad USBDEV_NAME definition bugfix
	
2006-12-30 - 0.2.1

	- command send repeats aren't needed any more

2006-12-24 - 0.2.0

	- fixed some problems with switching
	- added temporary switching in second tab (GtkNotebook)
	- about dialog added

2006-12-23 - 0.1

	v0.1 is out.. It can switch channels On/Off only..


powerSwitch.c
/*
 * powerSwitchGui.c
 * This file is part of gPowerSwitch
 *
 * Copyright (C) 2006 - Krzysiek Szczuka
 *
 * gPowerSwitch is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * gPowerSwitch is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with gPowerSwitch; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */



/*
	General Description:
	This program controls the PowerSwitch USB device using GTK+ GUI.
	It must be linked with libusb, a library for accessing the USB bus from
	Linux, FreeBSD, Mac OS X and other Unix operating systems. Libusb can be
	obtained from http://libusb.sourceforge.net/. Also it must be linked with
	GTK+ >= 2.0.
*/


#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <usb.h>				/* this is libusb, see http://libusb.sourceforge.net/ */
#include <gtk/gtk.h>

#include "powerSwitch.h"
//#include "support.h"
#include "psg-messages.h"


int
usbGetStringAscii (usb_dev_handle * dev, int index, int langid, char *buf,
				   int buflen)
{
	char buffer[256];
	int rval, i;

	if ((rval =
		 usb_control_msg (dev, USB_ENDPOINT_IN, USB_REQ_GET_DESCRIPTOR,
						  (USB_DT_STRING << 8) + index, langid, buffer,
						  sizeof (buffer), 1000)) < 0)
		return rval;
	if (buffer[1] != USB_DT_STRING)
		return 0;
	if ((unsigned char) buffer[0] < rval)
		rval = (unsigned char) buffer[0];
	rval /= 2;
	/* lossy conversion to ISO Latin1 */
	for (i = 1; i < rval; i++) {
		if (i > buflen)			/* destination buffer overflow */
			break;
		buf[i - 1] = buffer[2 * i];
		if (buffer[2 * i + 1] != 0)	/* outside of ISO Latin1 range */
			buf[i - 1] = '?';
	}
	buf[i - 1] = 0;
	return i - 1;
}


/* PowerSwitch uses the free shared default VID/PID. If you want to see an
 * example device lookup where an individually reserved PID is used, see our
 * RemoteSensor reference implementation.
 */


int
usbOpenDevice (usb_dev_handle ** device, int vendor, char *vendorName,
			   int product, char *productName)
{
	struct usb_bus *bus;
	struct usb_device *dev;
	usb_dev_handle *handle = NULL;
	int errorCode = USB_ERROR_NOTFOUND;
	static int didUsbInit = 0;

	if (!didUsbInit) {
		didUsbInit = 1;
		usb_init ();
	}
	usb_find_busses ();
	usb_find_devices ();
	for (bus = usb_get_busses (); bus; bus = bus->next) {
		for (dev = bus->devices; dev; dev = dev->next) {
			if (dev->descriptor.idVendor == vendor
				&& dev->descriptor.idProduct == product) {
				char string[256];
				int len;

				handle = usb_open (dev);	/* we need to open the device in order to query strings */
				if (!handle) {
					errorCode = USB_ERROR_ACCESS;
					fprintf (stderr,
							 "Warning: cannot open USB device: %s\n",
							 usb_strerror ());
					continue;
				}
				if (vendorName == NULL && productName == NULL) {	/* name does not matter */
					break;
				}
				/* now check whether the names match: */
				len =
					usbGetStringAscii (handle,
									   dev->descriptor.iManufacturer,
									   0x0409, string, sizeof (string));
				if (len < 0) {
					errorCode = USB_ERROR_IO;
					fprintf (stderr,
							 "Warning: cannot query manufacturer for device: %s\n",
							 usb_strerror ());
				} else {
					errorCode = USB_ERROR_NOTFOUND;
					/* fprintf(stderr, "seen device from vendor ->%s<-\n", string); */
					if (strcmp (string, vendorName) == 0) {
						len =
							usbGetStringAscii (handle,
											   dev->descriptor.iProduct,
											   0x0409, string,
											   sizeof (string));
						if (len < 0) {
							errorCode = USB_ERROR_IO;
							fprintf (stderr,
									 "Warning: cannot query product for device: %s\n",
									 usb_strerror ());
						} else {
							errorCode = USB_ERROR_NOTFOUND;
							/* fprintf(stderr, "seen product ->%s<-\n", string); */
							if (strcmp (string, productName) == 0)
								break;
						}
					}
				}
				usb_close (handle);
				handle = NULL;
			}
		}
		if (handle)
			break;
	}
	if (handle != NULL) {
		errorCode = 0;
		*device = handle;
	}
	return errorCode;
}


void
usbSend (int pscmd, int duration, int port)
{
	if( !handle )
		return;
	
	int duration2 = duration / 0.2 + 0.5;

	if (duration2 > 255)
		duration2 = 255;

	nBytes =
		usb_control_msg (handle,
						 USB_TYPE_VENDOR | USB_RECIP_DEVICE |
						 USB_ENDPOINT_IN, pscmd, duration2, port,
						 (char *) buffer, sizeof (buffer), 5000);
}


void
usbUpdate (void)
{
	if( !handle )
		return;
	
	nBytes =
		usb_control_msg (handle,
						 USB_TYPE_VENDOR | USB_RECIP_DEVICE |
						 USB_ENDPOINT_IN, PSCMD_GET, 0, 0,
						 (char *) buffer, sizeof (buffer), 5000);

	if (nBytes < 2) {
		if (nBytes < 0) {
			psg_message_dialog (GTK_MESSAGE_ERROR, "USB error: %s\n",
								usb_strerror ());
			gtk_exit (1);
		}
		fprintf (stderr, "only %d bytes status received\n", nBytes);
	}
}

void
usbTest (void)
{
	if( !handle )
		return;
	
	int i, v, r;

// The test consists of writing 1000 random numbers to the device and checking
// the echo. This should discover systematic bit errors (e.g. in bit stuffing).

	for (i = 0; i < 1000; i++) {
		v = rand () & 0xffff;
		nBytes =
			usb_control_msg (handle,
							 USB_TYPE_VENDOR | USB_RECIP_DEVICE |
							 USB_ENDPOINT_IN, PSCMD_ECHO, v, 0,
							 (char *) buffer, sizeof (buffer), 5000);
		if (nBytes < 2) {
			if (nBytes < 0) {
				fprintf (stderr, "USB error: %s\n", usb_strerror ());
				gtk_exit (1);
			}
			fprintf (stderr, "only %d bytes received in iteration %d\n",
					 nBytes, i);
		}
		r = buffer[0] | (buffer[1] << 8);
		if (r != v) {
			fprintf (stderr,
					 "data error: received 0x%x instead of 0x%x in iteration %d\n",
					 r, v, i);
		}
	}
	printf ("test succeeded\n");
}


powerSwitch.h
/*
 * powerSwitchGui.h
 * This file is part of powerSwitchGui
 *
 * Copyright (C) 2006 - Krzysiek Szczuka
 *
 * powerSwitchGui is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * powerSwitchGui is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with powerSwitchGui; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */


#ifndef _POWERSWITCH_GUI
#define _POWERSWITCH_GUI

#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif
#include <usb.h>



#define USBDEV_SHARED_VENDOR    0x16C0	/* VOTI */
#define USBDEV_SHARED_PRODUCT   0x05DC	/* Obdev's free shared PID */
/* Use obdev's generic shared VID/PID pair and follow the rules outlined
 * in firmware/usbdrv/USBID-License.txt.
 */

#define PSG_ACTIVE(v) (!v)
// definition for quick change of active state, because on my EvBoard active 
// is zero..


#define PSCMD_ECHO  0
#define PSCMD_GET   1
#define PSCMD_ON    2
#define PSCMD_OFF   3
/* These are the vendor specific SETUP commands implemented by our USB device */

#define USB_ERROR_NOTFOUND  1
#define USB_ERROR_ACCESS    2
#define USB_ERROR_IO        3


extern usb_dev_handle *handle;
extern unsigned char buffer[8];
extern int nBytes;



int usbGetStringAscii (usb_dev_handle * dev, int index, int langid,
					   char *buf, int buflen);


int usbOpenDevice (usb_dev_handle ** device, int vendor, char *vendorName,
				   int product, char *productName);


void usbSend (int pscmd, int duration, int port);


void usbUpdate (void);


void usbTest (void);


#endif


psg-messages.c
/*
 * psg-messages.c
 * This file is part of gPowreSwitch
 *
 * Copyright (C) 2006, 2007 - Krzysiek Szczuka
 *
 * gPowreSwitch is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * gPowreSwitch is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with gPowreSwitch; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */
 
 

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <gtk/gtk.h>

#include "psg-messages.h"
#include "powerSwitch.h"
//#include "support.h"



GtkWidget *
psg_message_dialog (GtkMessageType type, gchar * format, ...)
{

	va_list args;
	gchar *msg;

	g_return_val_if_fail (format != NULL, NULL);

	va_start (args, format);
	msg = g_strdup_vprintf (format, args);
	va_end (args);

	GtkWidget *dialog;

	dialog =
		gtk_message_dialog_new (NULL,
								GTK_DIALOG_MODAL |
								GTK_DIALOG_DESTROY_WITH_PARENT, type,
								GTK_BUTTONS_OK, msg);
	gtk_dialog_run (GTK_DIALOG (dialog));
	gtk_widget_destroy (dialog);
	g_free (msg);
	return dialog;
}


main.c
/*
 * main.c
 * This file is part of gPowerSwitch
 *
 * gPowerSwitch is GTK+2.0 based frontend for OBDEV's PowerSwitch.
 *  More info about PowerSwitch at: 
 *  http://www.obdev.at/products/avrusb/index.html
 *
 * Copyright (C) 2006 - Krzysiek Szczuka <kszczuka@gmail.com>
 *
 * gPowerSwitch is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * gPowerSwitch is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with gPowerSwitch; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, 
 * Boston, MA  02110-1301  USA
 */



#ifdef HAVE_CONFIG_H
#  include <config.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <usb.h>				/* this is libusb, see http://libusb.sourceforge.net/ */
#include <gtk/gtk.h>
#include <glade/glade-xml.h>

#include "powerSwitch.h"
#include "psg-messages.h"
#include "psg-callbacks.h"

usb_dev_handle *handle = NULL;
unsigned char buffer[8];
int nBytes;


GladeXML *xml = NULL;



static void
update_toggles (GtkWidget * dialog)
{
	usbUpdate ();
	
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c0")),
								  ((buffer[0] & (1 << 0)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c1")),
								  ((buffer[0] & (1 << 1)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c2")),
								  ((buffer[0] & (1 << 2)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c3")),
								  ((buffer[0] & (1 << 3)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c4")),
								  ((buffer[0] & (1 << 4)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c5")),
								  ((buffer[0] & (1 << 5)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c6")),
								  ((buffer[0] & (1 << 6)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
	gtk_toggle_button_set_active (GTK_TOGGLE_BUTTON
								  (glade_xml_get_widget (xml, "c7")),
								  ((buffer[0] & (1 << 7)) ?
								   PSG_ACTIVE (TRUE)
								   : PSG_ACTIVE (FALSE)));
}


#ifdef G_OS_WIN32
gchar *package_prefix;
gchar *package_datadir;
#endif


int
main (int argc, char *argv[])
{
	GtkWidget *dialog = NULL;
#ifdef G_OS_WIN32
	gchar *temp;

	package_prefix =
		g_win32_get_package_installation_directory (NULL, NULL);
	package_datadir = g_strdup_printf ("%s%s", package_prefix, "/share");
#endif

#ifdef ENABLE_NLS
#ifdef G_OS_WIN32
	temp = g_strdup_printf ("%s%s", package_prefix, "/lib/locale");
	bindtextdomain (GETTEXT_PACKAGE, temp);
	g_free (temp);
#else
	bindtextdomain (GETTEXT_PACKAGE, PACKAGE_LOCALE_DIR);
#endif
	bind_textdomain_codeset (GETTEXT_PACKAGE, "UTF-8");
	textdomain (GETTEXT_PACKAGE);
#endif
	
	gtk_set_locale ();
	gtk_init (&argc, &argv);
	
#ifdef G_OS_WIN32
	temp = g_strdup_printf ("%s%s", package_datadir, "/pixmaps");
	//add_pixmap_directory (temp);
	g_free (temp);
#else
	//add_pixmap_directory (PACKAGE_DATA_DIR "/" PACKAGE "/pixmaps");
#endif

	//dialog = create_dialog ();
	xml = glade_xml_new (GLADE_FILE, NULL, NULL);
	if( !xml ) {
	#ifdef G_OS_WIN32
		g_free (package_prefix);
		g_free (package_datadir);
	#endif
		return 1;
	}
	connect_signals ();
	dialog = glade_xml_get_widget (xml, "dialog");
	
	gtk_window_set_title (GTK_WINDOW (dialog), PACKAGE " v" VERSION);

	usb_init ();
	if (usbOpenDevice
		(&handle, USBDEV_SHARED_VENDOR, /*NULL */USBDEV_VENDOR_NAME  ,
		 USBDEV_SHARED_PRODUCT, /*NULL */ USBDEV_NAME ) != 0) {
		psg_message_dialog (GTK_MESSAGE_WARNING,
		//fprintf (stderr,
				 "Could not find USB device \"%s\" with vid=0x%x pid=0x%x\n",
				 USBDEV_NAME, USBDEV_SHARED_VENDOR, USBDEV_SHARED_PRODUCT);
		//exit (1);
	}
// We have searched all devices on all busses for our USB device above. Now
// try to open it and perform the vendor specific control operations for the
// function requested by the user.


	gtk_widget_show (dialog);
	update_toggles (dialog);

	gtk_main ();
	usb_close (handle);
#ifdef G_OS_WIN32
	g_free (package_prefix);
	g_free (package_datadir);
#endif
	return 0;
}

/*

int main(int argc, char **argv)
{

    if(strcmp(argv[1], "test") == 0){
        int i, v, r;
// The test consists of writing 1000 random numbers to the device and checking
// the echo. This should discover systematic bit errors (e.g. in bit stuffing).
 
        for(i=0;i<1000;i++){
            v = rand() & 0xffff;
            nBytes = usb_control_msg(handle, USB_TYPE_VENDOR | USB_RECIP_DEVICE | USB_ENDPOINT_IN, PSCMD_ECHO, v, 0, (char *)buffer, sizeof(buffer), 5000);
            if(nBytes < 2){
                if(nBytes < 0)
                    fprintf(stderr, "USB error: %s\n", usb_strerror());
                fprintf(stderr, "only %d bytes received in iteration %d\n", nBytes, i);
                exit(1);
            }
            r = buffer[0] | (buffer[1] << 8);
            if(r != v){
                fprintf(stderr, "data error: received 0x%x instead of 0x%x in iteration %d\n", r, v, i);
                exit(1);
            }
        }
        printf("test succeeded\n");
    }else if(strcmp(argv[1], "status") == 0){
        int i;
        nBytes = usb_control_msg(handle, USB_TYPE_VENDOR | USB_RECIP_DEVICE | USB_ENDPOINT_IN, PSCMD_GET, 0, 0, (char *)buffer, sizeof(buffer), 5000);
        if(nBytes < 2){
            if(nBytes < 0)
                fprintf(stderr, "USB error: %s\n", usb_strerror());
            fprintf(stderr, "only %d bytes status received\n", nBytes);
            exit(1);
        }
        for(i=0;i<8;i++){
            int isOn = buffer[0] & (1 << i);
            int isInv = buffer[1] & (1 << i);
            printf("port %d: %s%s\n", i, isOn ? "on" : "off", isInv ? (isOn ? " / pulse off" : " / pulse on") : "");
        }
    }else{
        int port, duration = 0;
        if(argc < 3){
            usage(argv[0]);
            exit(1);
        }
        port = atoi(argv[2]);
        if(argc > 3){
            if((duration = (int)(atof(argv[3]) / 0.2 + 0.5)) > 255)
                duration = 255;
        }
        if(strcmp(argv[1], "on") == 0){
            nBytes = usb_control_msg(handle, USB_TYPE_VENDOR | USB_RECIP_DEVICE | USB_ENDPOINT_IN, PSCMD_ON, duration, port, (char *)buffer, sizeof(buffer), 5000);
        }else if(strcmp(argv[1], "off") == 0){
            nBytes = usb_control_msg(handle, USB_TYPE_VENDOR | USB_RECIP_DEVICE | USB_ENDPOINT_IN, PSCMD_OFF, duration, port, (char *)buffer, sizeof(buffer), 5000);
        }else{
            nBytes = 0;
            usage(argv[0]);
            exit(1);
        }
        if(nBytes < 0)
            fprintf(stderr, "USB error: %s\n", usb_strerror());
    }
    usb_close(handle);
    return 0;
}
*/
